comms/src/dispatcher.rs

```diff
@@ -0,0 +1,186 @@
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use derive_error::Error;
+use std::{collections::HashMap, error::Error, hash::Hash};
+
+#[derive(Debug, Error, Clone)]
+pub enum DispatchError {
+    /// A dispatch route was not defined for the specific message type
+    MessageHandlerUndefined,
+    #[error(msg_embedded, non_std, no_from)]
+    HandlerError(String),
+}
+
+/// The signature of a handler function
+type HandlerFunc<M, E> = fn(msg: M) -> Result<(), E>;
+
+/// The trait bound for type parameter K on the dispatcher i.e the route key.
+/// i.e Must be Eq + Hash + {able to be sent across threads} +
+/// 'static (all references must live as long as the program BUT there are no references
+/// so therefore this is simply to satisfy the closure on `thread::spawn`)
+/// This simply saves us from having to put all these trait bounds whenever we
+/// want to apply these trait bounds (like a type alias).
+pub trait DispatchableKey: Eq + Hash + Send + 'static {}
+/// Implement this trait for all types which satisfy it's trait bounds.
+impl<T> DispatchableKey for T where T: Eq + Hash + Send + 'static {}
+
+/// A message type resolver. The resolver is called with the dispatched message.
+/// The resolver should then decide which dispatch key should be used.
+pub trait DispatchResolver<K, M>: Send + 'static
+where K: DispatchableKey
+{
+    fn resolve(&self, msg: &M) -> Result<K, DispatchError>;
+}
+
+/// Dispatcher pattern. Links handler function to "keys" which are resolved
+/// be a given type implementing [DispatchResolver].
+#[derive(Clone, Debug)]
+pub struct Dispatcher<K, M, E, R>
+where K: DispatchableKey
+{
+    handlers: HashMap<K, HandlerFunc<M, E>>,
+    catch_all: Option<HandlerFunc<M, E>>,
+    resolver: R,
+}
+
+impl<K, M, E, R> Dispatcher<K, M, E, R>
+where
+    K: DispatchableKey,
+    R: DispatchResolver<K, M>,
+    E: Error,
+{
+    /// Construct a new MessageDispatcher with no defined dispatch routes
+    pub fn new(resolver: R) -> Dispatcher<K, M, E, R> {
+        Dispatcher {
+            handlers: HashMap::new(),
+            resolver,
+            catch_all: None,
+        }
+    }
+
+    /// This function allows a new dispatch route to be specified and added to the handlers, all received messaged that
+    /// are of the dispatch type will be routed to the specified handler_function
+    pub fn route(mut self, path_key: K, handler: HandlerFunc<M, E>) -> Self {
+        self.handlers.insert(path_key, handler);
+        self
+    }
+
+    pub fn catch_all(mut self, handler: HandlerFunc<M, E>) -> Self {
+        self.catch_all = Some(handler);
+        self
+    }
+
+    /// This function can be used to forward a message to the correct function handler
+    pub fn dispatch(&self, msg: M) -> Result<(), DispatchError> {
+        let route_type = self.resolver.resolve(&msg)?;
+        match self.handlers.get(&route_type) {
```

Comment:
Could not use `self.handler.get(..).or(self.catch_all).and_then(..)...` because `handler.get` returns a reference, `catch_all` is owned and `or`'s signature is `fn(Option<T>) -> Option<T>` (same idea for `or_else`)

Reply:
`Option::as_ref` 🎉